dot/sync/outliers.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2019 ChainSafe Systems (ON) Corp.
+// This file is part of gossamer.
+//
+// The gossamer library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The gossamer library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+
+package sync
+
+import "math/big"
+
+//	removeOutlier removes the outlier from the slice
+//  Explanation:
+// 	IQR outlier detection
+// 	Q25 = 25th_percentile
+// 	Q75 = 75th_percentile
+// 	IQR = Q75 - Q25         // inter-quartile range
+// 	If x >  Q75  + 1.5 * IQR or  x   < Q25 - 1.5 * IQR THEN  x is a mild outlier
+// 	If x >  Q75  + 3.0 * IQR or  x   < Q25 – 3.0 * IQR THEN  x is a extreme outlier
+// Ref: http://www.mathwords.com/o/outlier.htm
+//
+// returns: sum of all the non-outliers elements
+func removeOutlier(sortedArr []*big.Int) (*big.Int, int64) {
+	length := len(sortedArr)
+
+	switch length {
+	case 0:
+		return big.NewInt(0), 0
+	case 1:
+		return sortedArr[0], 1
+	case 2:
+		return big.NewInt(0).Add(sortedArr[0], sortedArr[1]), 2
+	}
+
+	half := length / 2
+	data1 := sortedArr[:half]
+	var data2 []*big.Int
+
+	if length%2 == 0 {
+		data2 = sortedArr[half:]
+	} else {
+		data2 = sortedArr[half+1:]
+	}
+
+	q1 := getMedian(data1)
+	q3 := getMedian(data2)
+
+	iqr := big.NewInt(0).Sub(q3, q1)
+	iqr1_5 := big.NewInt(0).Mul(iqr, big.NewInt(2)) //instead of 1.5 it is 2.0 due to the rounding
```

Comment:
this seems like it'll change the math of the equation - maybe you can use `big.Float` here?

Reply:
@noot not really needed since the IQR range is calculated in the range from 1.5 to 3. Where 1.5 is a more compact range and 3 is a larger range for an element to be named an outlier. So, 2 is, in my opinion, a sweet spot for us.